frontend/javascripts/oxalis/controller/mesh_controller.ts

```diff
@@ -0,0 +1,167 @@
+    mesh.receiveShadow = true;
+    const tweenAnimation = new TWEEN.Tween({
+      opacity: 0,
+    });
+    tweenAnimation
+      .to(
+        {
+          opacity: 1,
+        },
+        500,
+      )
+      .onUpdate(function onUpdate() {
+        // @ts-expect-error ts-migrate(2683) FIXME: 'this' implicitly has type 'any' because it does n... Remove this comment to see the full error message
+        meshMaterial.opacity = this.opacity;
+        app.vent.trigger("rerender");
+      })
+      .start();
+    return mesh;
+  }
+
+  addIsosurfaceFromGeometry(
+    geometry: THREE.BufferGeometry,
+    segmentationId: number,
+    offset: Vector3 | null = null,
+    scale: Vector3 | null = null,
+    lod: number,
+    layerName: string,
+  ): void {
+    if (this.isosurfacesGroupsPerSegmentationId[layerName] == null) {
+      this.isosurfacesGroupsPerSegmentationId[layerName] = {};
+    }
+    if (this.isosurfacesGroupsPerSegmentationId[layerName][segmentationId] == null) {
+      this.isosurfacesGroupsPerSegmentationId[layerName][segmentationId] = {};
+    }
+    if (this.isosurfacesGroupsPerSegmentationId[layerName][segmentationId][lod] == null) {
+      const newGroup = new THREE.Group();
+      this.isosurfacesGroupsPerSegmentationId[layerName][segmentationId][lod] = newGroup;
+      if (lod === NO_LOD_MESH_INDEX) {
+        this.isosurfacesLODRootGroup.addNoLODSupportedMesh(newGroup);
+      } else {
+        this.isosurfacesLODRootGroup.addLODMesh(newGroup, lod);
+      }
+      // @ts-ignore
+      newGroup.cellId = segmentationId;
+      if (scale != null) {
+        newGroup.scale.copy(new THREE.Vector3(...scale));
+      }
+    }
+    const mesh = this.constructIsosurfaceMesh(segmentationId, geometry);
+    if (offset) {
+      mesh.translateX(offset[0]);
+      mesh.translateY(offset[1]);
+      mesh.translateZ(offset[2]);
+    }
+
+    this.isosurfacesGroupsPerSegmentationId[layerName][segmentationId][lod].add(mesh);
+  }
+
+  removeIsosurfaceById(segmentationId: number, layerName: string): void {
+    if (this.isosurfacesGroupsPerSegmentationId[layerName] == null) {
+      return;
+    }
+    if (this.isosurfacesGroupsPerSegmentationId[layerName][segmentationId] == null) {
+      return;
+    }
+    _.forEach(
+      this.isosurfacesGroupsPerSegmentationId[layerName][segmentationId],
+      (meshGroup, lod) => {
+        const lodNumber = parseInt(lod);
+        if (lodNumber !== NO_LOD_MESH_INDEX) {
+          this.isosurfacesLODRootGroup.removeLODMesh(meshGroup, lodNumber);
+        } else {
+          this.isosurfacesLODRootGroup.removeNoLODSupportedMesh(meshGroup);
+        }
+      },
+    );
+    delete this.isosurfacesGroupsPerSegmentationId[layerName][segmentationId];
+  }
+
+  getIsosurfaceGeometryInBestLOD(segmentId: number, layerName: string): THREE.Group {
+    const bestLod = Math.min(
+      ...Object.keys(this.isosurfacesGroupsPerSegmentationId[layerName][segmentId]).map((lodVal) =>
+        parseInt(lodVal),
```

Comment:
I think, the case in which the look ups would fail are unexpected (e.g., the UI shouldn't expose setIsosurfaceVisibility if there is no corresponding mesh) which is why they should fail. Adding a defensive check only to throw another error doesn't feel valuable to me. Doing a noop could be the alternative, but I'd rather throw an exception here. Since this will likely be in an eventhandler, wk shouldn't crash anyway.